### PR TITLE
fix: [CDS-41281]: refactored /metadata API in servicenow

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- Changes json parsing to also parse schema information.
- Corrected implementation of parsing `required`, `custom` , `key`, `name`, schema object, `allowedValues` in helper function
- Added correct typeStr for `option` type of field in servicenow
- Sys_Id is marked as non mandatory, as we have delegate logic to get sys_id in update step.
- Added a field 'multiText' which will help UI to show fields as expandable fields. 
- a new action type introduced which handled parsing on manager side, and unused faulty parsing logic in beans deleted.
- createMetadata and metadata API use the new action when FF is on and delegates are upgraded, but resolve to current behaviour when FF is off or FF is on with delegates old
- Testing
  - Tested Required fields ; Custom Fields ; MultiText fields; Unknown field;  Boolean field ;  Integer field ; Date time field ;String field; Choice fields to be populating correctly in new implementation
  - testing metadata API call for regression testing in Snow create / update -> UI need to handle unknown and boolean type of fields. Rest looks good. ([Handling unknown and boolean type field](https://harness.atlassian.net/browse/CDS-69975?atlOrigin=eyJpIjoiOTU3MGQ2ZGZjM2I0NGEyYzg4ZGJjZDk1ODE3YTMxMjQiLCJwIjoiamlyYS1zbGFjay1pbnQifQ))
  - affect on createMetadata call : just one multiText field added with value as false always
  - Checked whether new metadata call has all info used by UI in /createMetadata call in approval, so that /metadata can be used here.
  - delegate backward compatibility checking
  - FF on ;with delegates upgraded; (with both endpoints)
  - FF on; with delegates old; (with both endpoints)
  - FF off; (with both endpoints)
  - createMetadata API with FF on and delegate upgraded in breaking the FE

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48275)

### CODE REVIEW

File: 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java [hunk 1]
```diff
@@ -32,18 +34,24 @@ public class ServiceNowFieldSchemaNG {
   @NotNull String typeStr;
   @NotNull ServiceNowFieldTypeNG type;
   String customType;
+  boolean isMultiText;
```
Comment (naman-goenka) [inline 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java:37]: `isMultilineText`

File: 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java [hunk 2]
```diff
@@ -32,18 +34,24 @@ public class ServiceNowFieldSchemaNG {
...
 
   public ServiceNowFieldSchemaNG(JsonNode node) {
     this.typeStr = JsonNodeUtils.getString(node, "type", "string");
```
Comment (naman-goenka) [inline 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java:40]: Move to a utils class

File: 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldTypeNG.java [hunk 3]
```diff
@@ -25,7 +25,7 @@ public enum ServiceNowFieldTypeNG {
   @JsonProperty("integer") INTEGER(Collections.singletonList("integer")),
   @JsonProperty("boolean") BOOLEAN(Collections.singletonList("boolean")),
   @JsonProperty("string") STRING(Collections.singletonList("string")),
-  @JsonProperty("option") OPTION(Collections.singletonList("option")),
+  @JsonProperty("option") OPTION(Arrays.asList("option", "choice")),
   @JsonProperty("unknown") UNKNOWN(Collections.emptyList());
 
   @Getter private List<String> snowInternalTypes;
```
Comment (naman-goenka) [inline 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldTypeNG.java:31]: Add logs when inavalid typeStr is given for debugging purpose as we are setting it as empty

File: 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java (hunk 1 above)
Comment (naman-goenka) [reply 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java:37]: done

File: 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java (hunk 2 above)
Comment (naman-goenka) [reply 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldSchemaNG.java:40]: done

File: 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldTypeNG.java (hunk 3 above)
Comment (naman-goenka) [reply 970-api-services-beans/src/main/java/io/harness/servicenow/ServiceNowFieldTypeNG.java:31]: done

**Review by puthrayaharness — DISMISSED** [review]

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger publish-delegate

**Comment by naman-goenka** [comment]
trigger uts

**Review by vivekkarnatiharness — APPROVED** [review]

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger ut3

**Review by puthrayaharness — APPROVED** [review]

**Pull request merged**